### PR TITLE
9 executive logic overarching logic

### PR DESCRIPTION
**Pull request by Wilson-Orlando**
- Implemented `ExecutiveLogic` and `Wheel`
- Reimplemented `Server` and `Client`
- Small fixes to other files
- Verified that everything can run through a small portion of a WoJ game

This turned into a huge PR, my bad. If you pull this branch and run `main.py` and `client.py` in separate terminals, you should be able to play through part of the first round of a WoJ game.

### CODE REVIEW

**Comment by Wilson-Orlando** [comment]
This is in dire need of unit tests, by the way, but Exec/Server/Client are so interconnected that I was having trouble mapping tests out. By running the game with `main.py` and` client.py`, I was able to catch a few errors, so you can get through a few questions in the first round before the game stalls. Testing/debugging the whole thing may become a separate PR.

File: executive_logic/executive_logic.py [hunk 1]
```diff
@@ -1,36 +1,204 @@
...
+
+        # Spin wheel
+        if self.num_spins >= MAX_SPINS:
+            return
+        wheel_result = self.__spin_wheel()
+        #self.__query_server(MessageType.SPIN_RESULT, [wheel_result])  # TODO: Implement, so that user can see what sector they spun
+
+        # If result is a Jeopardy category, call execute_category
+        if self.wheel.is_jeopardy_category(wheel_result):
+            self.__execute_category(wheel_result, curr_player_id, round_num)
```
Comment (robyntorregrosa) [inline executive_logic/executive_logic.py:70]: This should continue if the player gets it right. I think you need to return `is_correct` from __execute_category and continue until the category is empty or the player `is_correct` is False.

File: server.py [hunk 2]
```diff
@@ -13,183 +12,193 @@
...
+    EMPTY = 0  # No message.
+    # Request  Args: 	[]
+    # Response Args:	[]
 
-	def host_game(self):
-		server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)  # setup server socket
-		server.bind((self.host_ip, self.port))
-		logging.info("server starting a game, listening for players on port: " + str(self.port))
-		server.listen(NUM_PLAYERS - 1)  # listen for 2 connections. 3 player game
+    PLAYER_ID = 1  # Called by server the first time it connects to a client. Assigns that client its unique player ID.
```
Comment (robyntorregrosa) [inline server.py:38]: I noticed `player_id` is 1+ in the gameserver, but 0, 1 or 2 in the executive logic. So the first player to join gets a message the their `player_id` is 1. We should make sure these match. I think it's good to start with player 0.

File: executive_logic/executive_logic.py [hunk 3]
```diff
@@ -1,36 +1,204 @@
...
-    def selectRandOpponent(self):
-        return "Opponent"
+    def __update_scores_tokens_spins(self):
+        """
+        Called after a game value is updated that should be reflected in the UI display.
+        Sends server the current player scores, player tokens, and number of spins.
+        :return: void
+        """
+        self.__query_server(MessageType.UPDATE_SCORES,
+                            [self.score_keeper.get_scores(), self.score_keeper.get_tokens(), self.num_spins])
```
Comment (robyntorregrosa) [inline executive_logic/executive_logic.py:179]: I think we should send the number of spins remaining, that way the Client/UI doesn't also have to keep up with MAX_SPINS.

File: executive_logic/executive_logic.py [hunk 4]
```diff
@@ -1,36 +1,204 @@
...
+
+        elif wheel_result == Sector.PLAYERS_CHOICE:
+            self.__query_server(MessageType.PLAYERS_CHOICE, [curr_player_id, round_num])
+            _, [chosen_category] = self.query_response
+            self.__execute_category(chosen_category, curr_player_id, round_num)
+
+        elif wheel_result == Sector.OPPONENTS_CHOICE:
+            self.__query_server(MessageType.OPPONENTS_CHOICE, [curr_player_id, round_num])
+            _, [chosen_category] = self.query_response
+            self.__execute_category(chosen_category, curr_player_id, round_num)
```
Comment (robyntorregrosa) [inline executive_logic/executive_logic.py:99]: TODO: have the exec logic pick a random opponent that is NOT `curr_player_id`

File: client.py [hunk 5]
```diff
@@ -13,113 +15,192 @@
...
-		# 		# elif parse_request[0] == '' # some other command
-		# 		else:
-		# 			self.refresh_display()
-
-	def refresh_score(self, score):
-		"""
+    # default constructor
+    def __init__(self):
+        self.player_id = None
+        self.whose_turn = 1  # TODO: Unused. Currently, each client doesn't know whose turn it is; they're only notified when server asks them something
```
Comment (robyntorregrosa) [inline client.py:24]: Go ahead and remove `whose_turn`

File: executive_logic/executive_logic.py [hunk 6]
```diff
@@ -1,36 +1,204 @@
...
 class ExecutiveLogic:
-    def __init__(self):
-        self.num_spins = 1
-        self.round = 1
-        pass
+    def __init__(self, srv_ip, srv_port):
+        self.board = Board()  # Board object
+        self.score_keeper = ScoreKeeper()  # ScoreKeeper object
+        self.wheel = Wheel(["food", "politics"])  # Wheel object (TODO: These categories should be specified somewhere else, right?)
+        self.ui = UserInterface()  # UI Object
```
Comment (robyntorregrosa) [inline executive_logic/executive_logic.py:17]: You can remove `self.ui` here

File: executive_logic/executive_logic.py [hunk 7]
```diff
@@ -1,36 +1,204 @@
...
+
+    def run_game(self):
+        """
+        Main game loop. Controls logic through the flow of the entire game.
+        :return:
+        """
+        self.is_game_running = True
+
+        for round_num in [1, 2]:
+            self.board.reset_board(round_num)
```
Comment (robyntorregrosa) [inline executive_logic/executive_logic.py:34]: For the bankrupt, does it bankrupt you just for the round? If so, we have to make a function to store the round 1 scores in the scorekeeper and reset the scores for round 2.

File: client.py [hunk 8]
```diff
@@ -13,113 +15,192 @@
...
+        :param prompt_list: List of strings to pick from (e.g. a list of answers to a Jeopardy question
+        :return: (string) The answer that the user selected
+        """
+        # Print out the options in the list
+        for i in range(0, len(prompt_list)):
+            print("----- " + str(i) + ") " + str(prompt_list[i]))
+        # Ask user to select an option from the list
+        selected_index = None
+        while selected_index not in range(0, len(prompt_list)):
+            max_index = str(len(prompt_list))
```
Comment (robyntorregrosa) [inline client.py:175]: Since you're starting index from 0,
```suggestion
            max_index = str(len(prompt_list)-1)
```

**Review by robyntorregrosa — APPROVED** [review]
This is great! Thanks for making those changes.

**Pull request merged**